### PR TITLE
Heapster Librato sink

### PR DESCRIPTION
**Pull request by johanneswuerbach**
Sink for https://www.librato.com/

### CODE REVIEW

**Comment by johanneswuerbach** [comment]
I signed it!

**Comment by dieterdemeyer** [comment]
@johanneswuerbach we could also use this sink if it gets merged.
Do you happen to have a timeline on this ? If I can help in any way, please let me know.

**Comment by johanneswuerbach** [comment]
@dieterdemeyer I haven't gotten any feedback from any k8s maintainer yet, so hard to say.

We are currently building a custom heapster from this branch using:
`OVERRIDE_IMAGE_NAME=org/custom-heapster:test-7 make container push-override`

**Comment by dieterdemeyer** [comment]
@johanneswuerbach how exactly should I use that ? I'm just starting with Kubernetes, so bear with me..

**Comment by johanneswuerbach** [comment]
@dieterdemeyer you can use a config similar to the once you can find here https://github.com/kubernetes/kubernetes/blob/master/cluster/addons/cluster-monitoring/influxdb/heapster-controller.yaml, but instead of the official image you can use your custom own and instead of influxdb, you can use the librato sink

**Comment by dieterdemeyer** [comment]
@johanneswuerbach Thanks, I'll give it a try..

File: Makefile [hunk 1]
```diff
@@ -89,7 +89,7 @@ container:
 
 	docker build --pull -t $(PREFIX)/heapster-$(ARCH):$(VERSION) $(TEMP_DIR)
 ifneq ($(OVERRIDE_IMAGE_NAME),)
-	docker tag -f $(PREFIX)/heapster-$(ARCH):$(VERSION) $(OVERRIDE_IMAGE_NAME)
+	docker tag $(PREFIX)/heapster-$(ARCH):$(VERSION) $(OVERRIDE_IMAGE_NAME)
```
Comment (DirectXMan12) [inline Makefile]: did you mean to change the Makefile?

File: common/librato/librato.go [hunk 2]
```diff
@@ -0,0 +1,134 @@
+// Copyright 2017 Google Inc. All Rights Reserved.
```
Comment (DirectXMan12) [inline common/librato/librato.go]: this doesn't need to be in common unless you also plan on writing an eventer sink.

**Comment by DirectXMan12** [comment]
Following up, the code looks decent.  It should be good to merge once we've confirmed that it passes tests and you fix the nits.

**Comment by DirectXMan12** [comment]
@k8s-bot ok to test

File: Makefile (hunk 1 above)
Comment (johanneswuerbach) [inline Makefile]: This line never worked for me as https://docs.docker.com/engine/reference/commandline/tag/ has no `-f` flag. I can send a separate PR if you prefer that.

File: common/librato/librato.go (hunk 2 above)
Comment (johanneswuerbach) [inline common/librato/librato.go]: I was planning to, but https://www.librato.com/docs/api/#annotations isn't ready for tags yet. I'll contact librato to find out a timeframe.

**Comment by johanneswuerbach** [comment]
Thanks @DirectXMan12 and no worries, I updated the sink owners file and left some comments. Let me know if anything else is required :-)

File: common/librato/librato.go (hunk 2 above)
Comment (DirectXMan12) [inline common/librato/librato.go]: ack, alright :-)

File: Makefile (hunk 1 above)
Comment (DirectXMan12) [inline Makefile]: huh, you appear to be correct, this should be fine.

**Comment by dieterdemeyer** [comment]
@johanneswuerbach I used the master branch of kubernetes/heapster and built a new image with the following command:

`
OVERRIDE_IMAGE_NAME=myorg/heapster:1.3.0-SNAPSHOT make container push-override
`

I'm trying to use the librato sink but I'm getting the following error message when the pod starts:
`
Failed to create sink: Sink not recognized: "librato
`

I'm using the following configuration in my container command:
`
--sink="librato:?username=<username>&token=<token>&prefix=k8s&tags=cluster"
`

Could you tell me what I'm doing wrong ?

Thanks in advance.

**Comment by dieterdemeyer** [comment]
@johanneswuerbach Never mind, just found the cause.
The documentation found [here](https://github.com/kubernetes/heapster/blob/master/docs/sink-configuration.md) describes quotes around the sink configuration but it actually shouldn't contain quotes.

**Comment by johanneswuerbach** [comment]
Thanks @dieterdemeyer would you mind sending a PR removing the quotes? That would be great 👍

**Comment by dieterdemeyer** [comment]
@johanneswuerbach Sure, will do..
Did you test the Librato sink with real live data ? I have some weird values in the metric k8s.cpu.usage...
Are you available in other channels like the Kubernetes slack instance for a small talk about this ?

**Comment by johanneswuerbach** [comment]
Yes we are currently using it with 3 clusters, but I didn't check all the values yet. Not in the slack yet, but will join once again on a computer

File: Makefile (hunk 1 above)
Comment (piosz) [inline Makefile]: FYI, this broke e2e tests
```
docker tag gcr.io/google_containers/heapster-amd64:v1.3.0-beta.1 heapster:e2e_test
Error response from daemon: Conflict: Tag heapster:e2e_test is already set to image 1b32c9c98284, if you want to replace it, please use -f option
Makefile:68: recipe for target 'container' failed
make[1]: *** [container] Error 1
make[1]: Leaving directory '/go/src/k8s.io/heapster
```
https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/heapster/1572/pull-heapster-e2e/238/

I'm going to revert this change (only Makefile part).

Comment (piosz) [inline Makefile]: ref #1572

**Comment by mashayev** [comment]
Hii @johanneswuerbach @dieterdemeyer I got the same issue with `Failed to create sink: Sink not recognized: librato`

I removed the quotes from sink and it's the same issue. Maybe `heapster-amd64:v1.3.0-beta.1` image need to be updated?

```
apiVersion: extensions/v1beta1
kind: Deployment
metadata:
  name: heapster
  namespace: kube-system
spec:
  replicas: 1
  template:
    metadata:
      labels:
        task: monitoring
        k8s-app: heapster
    spec:
      containers:
      - name: heapster
        image: gcr.io/google_containers/heapster-amd64:v1.3.0-beta.1
        imagePullPolicy: Always
        command:
        - /heapster
        - --source=kubernetes:https://kubernetes.default
        - --sink=librato:?username=xyz&token=secret&prefix=k8s&tags=cluster&tag_cluster=staging
        volumeMounts: 
          - name: ssl-certs
            mountPath: "/etc/ssl/certs/ca-certificates.crt"
            readOnly: true
      volumes: 
        - name: ssl-certs
          hostPath: 
            path: "/etc/pki/tls/certs/ca-bundle.crt"
```

**Comment by johanneswuerbach** [comment]
@mashayev beta.1 doesn't contain the librato sink, but the recently released https://github.com/kubernetes/heapster/releases/tag/v1.3.0 does. Could you double check?

**Comment by mashayev** [comment]
@johanneswuerbach Thanks for the help, it works :)

**Comment by mashayev** [comment]
@johanneswuerbach One more question about "The librato sink currently only works with accounts, which support tagged metrics."
If we are using source and not tags(We signed up on or before January 24th, 2017 12:00 PST), it means the sink wont work at all?

Any plan to support both??

**Comment by johanneswuerbach** [comment]
I'm only using tagged metrics and as far as I know source based metrics are also more or less deprecated. As far as I know you can just sent an email to librato support and your account will be whitelisted to use both.

**Comment by mashayev** [comment]
@johanneswuerbach Librato team have enabled the Tags for my account, the Sink is working now for me.
I have another question: is there any option to change the metrics name in the chart?
for example I get this name in the chart:
`i-1q2w3edfrtt.ip-1-1-1-1.ec2.internal.node.ip-1-1-1-1.beta.kubernetes.io_arch:amd64_beta.kubernetes.io_instance-type:m3.large_beta.kubernetes.io_os:linux_failure-domain.beta.kubernetes.io_region:us-east-1_failure-domain.beta.kubernetes.io_zone:us-east-1c_kubernetes.io_hostname:ip-1-1-1-1`

I prefer to get the name of one of the tags or combination of them for example. something like:
`<CLUSTER_REGION>.<ENVIRONMENT>.<POD_NAME>.<POD_NAMESPACE>`
Below is an example of how it works when I send the metrics with Collectd(I can control the way the name will be displayed), is it possible with Librato Sink?

image

Thanks for the help.

**Comment by johanneswuerbach** [comment]
Not sure, as far as I know the metric name is auto-generated from all received tags, but I guess you could ask librato support for clarification.

The sink is only forwarding each metric with all submitted tags to https://www.librato.com/docs/api/#create-a-measurement

**Comment by mashayev** [comment]
Hi @johanneswuerbach,
I wonder if it is possible to filter which metrics will be sent to Librato and which not. From what I see there is no such an option. Can you confirm?

**Comment by johanneswuerbach** [comment]
Yes, that is correct.
